features/nanostack/FEATURE_NANOSTACK/targets/TARGET_SL_RAIL/NanostackRfPhyEfr32.cpp

```diff
@@ -152,6 +186,76 @@ static int8_t rf_start_cca(uint8_t *data_ptr, uint16_t data_length, uint8_t tx_h
 /* Local function prototypes */
 static bool rail_checkAndSwitchChannel(uint8_t channel);
 
+static void rf_thread_loop(const void *arg)
+{
+    tr_debug("rf_thread_loop: starting (id: %d)\n", rf_thread_id);
+    for (;;) {
+        osEvent event = osSignalWait(0, osWaitForever);
+
+        if (event.status != osEventSignal) {
+            continue;
+        }
+
+        platform_enter_critical();
```

Comment:
Isn't this going to create the same issue with nanostack? You're calling nanostack (phy_rx_cb) from within a critical section... Not sure you can do mutex operations in there?

Reply:
Since what you're protecting is _head and _tail, I'd suggest caching their values into a local variable, to not run into a concurrent-update situation, and keep that critical section as short as possible.

Reply:
In Mbed OS the Nanostack's `platform_enter_critical()` is just a mutex.
See [here](https://github.com/ARMmbed/mbed-os/blob/master/features/FEATURE_COMMON_PAL/nanostack-hal-mbed-cmsis-rtos/arm_hal_interrupt.c#L28).

So it is a way of protecting Nanostack to call you when you are in a state that you cannot receive TX-start of any of this sort events.